Data_Engine/Compute/DomainTreeClusters.cs

```diff
@@ -0,0 +1,89 @@
+/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2020, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Data.Collections;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace BH.Engine.Data
+{
+    public static partial class Compute
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static List<List<T>> DomainTreeClusters<T>(this List<T> data, Func<T, DomainBox> toDomainBox, Func<DomainBox, DomainBox, bool> TreeEvaluator, Func<T, T, bool> itemEvaluator, int minPointCount = 1)
+        {
+            if (data.Count == 0)
+                return new List<List<T>>();
+
+            List<T> items = new List<T>(data);
+            List<bool> check = items.Select(x => false).ToList();
+            DomainTree<int> indexTree = Data.Create.DomainTree(items.Select((x, i) => Data.Create.DomainTreeLeaf(i, toDomainBox(x))));
```

Comment:
Wouldn't that be better to call `toDomainBox` only once (now it is called here and in line 64)? Not sure whether any improvement in practice, but assuming that the method can be as heavy as one can imagine, that could help avoiding overhead.

Reply:
Excellent point, can just pre-process them and store in a list 👍